packages/entities/src/lib/update.mutation.ts

```diff
@@ -174,3 +175,55 @@ export function upsertEntities<
     return addEntities(newEntities, options)(newState, context) as S;
   };
 }
+
+/**
+ *
+ * Merge entities that exists, add those who don't
+ * Make sure all entities have an id
+ *
+ * @example
+ *
+ * // single entity
+ * store.update(upsertEntities({ id: 1, completed: true }))
+ *
+ * // or multiple entities
+ * store.update(upsertEntities([{ id: 1, completed: true }, { id: 2, completed: true }]))
+ *
+ * // or using a custom ref
+ * store.update(upsertEntities([{ id: 1, open: true }], { ref: UIEntitiesRef }))
+ *
+ */
+export function upsertEntities<
+  S extends EntitiesState<Ref>,
+  Ref extends EntitiesRef = DefaultEntitiesRef
+>(
+  entities: OrArray<Partial<getEntityType<S, Ref>>>,
+  options: AddEntitiesOptions & BaseEntityOptions<Ref> = {}
+): Reducer<S> {
+  return function (state, context) {
+    const { prepend = false, ref = defaultEntitiesRef } = options;
+    const { entitiesKey, idsKey } = ref!;
+    const idKey = getIdKey<getIdType<S, Ref>>(context, ref);
+
+    const asObject = {} as Record<getIdType<S, Ref>, getEntityType<S, Ref>>;
+    const ids = [] as getIdType<S, Ref>;
+    for (const entity of coerceArray(entities)) {
```

Comment:
if `!coerceArray(entities).length` return the original state

Reply:
Fixed